WordPress/gutenberg
# Add URL Details endpoint to REST API to allow retrieval of info about a remote URL

Closes https://github.com/WordPress/gutenberg/issues/18047.

Adds a new REST API endpoint to grab details from a external URL. So now if I want to get the contents of the `title` tag on `https://wordpress.org` I can send a request to this API endpoint and it will return the value in the response for use in the Gutenberg UI.

Why might this be useful? See https://github.com/WordPress/gutenberg/pull/19387.

-----

~As part of https://github.com/WordPress/gutenberg/pull/17846 we need to display the contents of the `<title>` tag from the url the user has typed into the input field.~ This has now long since shipped. Nonetheless, it might be a nice enhancement to be able to show details about the remote URL that has been selected.

This PR addresses this by adding a new **experimental** custom Gutenberg REST API endpoint. This endpoint:

* makes a request for a remote URL.
* ~parses the response in a `DOMDocument`~ we should investigate this in future (see [here](https://github.com/google/web-stories-wp/blob/a190acea43fc8dcac937d95b114d61ef4cad6aad/includes/REST_API/Link_Controller.php))
* retrieves the `title` tag contents.

To prevent abuse, the endpoint is authenticated and requires the user to have the `edit_posts` capability to be able to make a request.

## How has this been tested?

- [x] Manual testing (see below)

## Automated Testing Instructions

`npm run test-php` should run the unit tests which reside in `phpunit/class-wp-rest-url-details-controller-test.php`.

## Manual Testing Instructions

* Check out this PR.
* Open the Gutenberg local testing env at `http://localhost:8889/wp-admin/` and login.
* Create a new Post.
* Choose one of the following options:

### Option 1

* Open devtools and in the `console` enter:

```
wp.apiFetch( { path: '/__experimental/url-details/?url=https://wordpress.org' } ).then( data => {
    console.log( data );
} );
```

You'll see the response returned there.

### Option 2
* Open devtools and in the `console` enter `wpApiSettings.nonce`. You should see a valid nonce returned as a string.
* Copy the nonce.
* In your browser goto: `http://localhost:8889/wp-json/__experimental/url-details/?url=https://wordpress.org&_wpnonce=%YOUR_NONCE_HERE%` - be sure to replace the nonce value with that copied from the previous step.
* You should see a valid REST API response containing the contents of the title tag from `wordpress.org` (ie: `"Blog Tool, Publishing Platform, and CMS \u2014 WordPress.org"`).



## Types of changes
New feature (non-breaking change which adds functionality).

## Checklist:
